src/mlpack/core/optimizers/cne/cne.hpp

```diff
@@ -0,0 +1,220 @@
+    const size_t maxGenerations = 5000,
+    const double mutationProb = 0.1,
+    const double mutationSize = 0.02,
+    const double selectPercent = 0.2,
+    const double finalValue = DBL_MIN,
+    const double fitnessHist = DBL_MIN);
+
+/**
+ * Optimize the given function using CNE. The given
+ * starting point will be modified to store the finishing point of the
+ * algorithm, and the final objective value is returned.
+ *
+ * @tparam DecomposableFunctionType Type of the function to be optimized.
+ * @param function Function to optimize.
+ * @param iterate Starting point (will be modified).
+ * @return Objective value of the final point.
+ */
+  template<typename DecomposableFunctionType>
+  double Optimize(DecomposableFunctionType& function, arma::mat& iterate);
+
+  //! Get the population size
+  size_t PopulationSize() const { return populationSize; }
+  //! Modify the population size
+  size_t& PopulationSize() { return populationSize; }
+
+  //! Get maximum number of generations
+  size_t MaxGenerations() const { return maxGenerations; }
+  //! Modify maximum number of generations
+  size_t& MaxGenerations() { return maxGenerations; }
+
+  //! Get the mutation probability
+  double MutationProbability() const { return mutationProb; }
+  //! Modify the mutation probability
+  double& MutationProbability() { return mutationProb; }
+
+  //! Get the mutation size
+  double MutationSize() const { return mutationSize; }
+  //! Modify the mutation size
+  double& MutationSize() { return mutationSize; }
+
+  //! Get the selection percentage
+  double SelectionPercentage() const { return selectPercent; }
+  //! Modify the selection percentage
+  double& SelectionPercentage() { return selectPercent; }
+
+  //! Get the final objective value
+  double FinalValue() const { return finalValue; }
+  //! Modify the final objective value
+  double& FinalValue() { return finalValue; }
+
+  //! Get the change in fitness history between generations
+  double FitnessHistory() const { return fitnessHist; }
+  //! Modify the termination criteria of change in fitness value
+  double& FitnessHistory() { return fitnessHist; }
+
+ private:
+  //! Reproduce candidates to create the next generation
+  void Reproduce();
+
+  //! Function to modify weights for the evolution of next generation
+  void Mutate();
+
+  /**
+   * Crossover parents and create new childs. 
+   * Two parents create two new childs.
+   *
+   * @param mom      First parent from the elite population
+   * @param dad      Second parent from the elite population
+   * @param dropout1 The place to delete the candidate of the present
+   *                 generation and place a child over there for the
+   *                 next generation
+   * @param dropout2 The place to delete the candidate of the present
+   *                 generation and place a child over there for the
+   *                 next generation
+   */
+  void Crossover(size_t mom, size_t dad, size_t dropout1, size_t dropout2);
```

Comment:
@zoq 
made all the parameters const in crossover.
and instead using 
```
    // Parents generates 2 childs replacing the droped out candidates.
    // The index of the candidates in the population is found first. 
    Crossover(index[mom], index[dad], index[i], index[i+1]);
```
and then 
```
//! Crossover parents to create new childs.
void CNE::Crossover(const size_t mom,
                    const size_t dad,
                    const size_t child1,
                    const size_t child2)
```

Just mentioning. I think the previous way of finding the index inside crossover was looking more beautiful and friendly to read. 

```
//! Crossover parents to create new childs.
void CNE::Crossover(size_t mom,
                    size_t dad,
                    size_t child1,
                    size_t child2)
{
  // Find the index of these candidates in the population matrix.
  mom = index[mom];
  dad = index[dad];
  child1 = index[child1];
  child2 = index[child2];
```